libs/ledger/src/upow/synergetic_executor.cpp

```diff
@@ -87,15 +111,34 @@ void SynergeticExecutor::Verify(WorkQueue &solutions, ProblemData const &problem
       contract->Attach(storage_);
       ContractContext ctx(&token_contract_, solution->address(), &storage_adapter, 0);
       contract->UpdateContractContext(ctx);
-      status = contract->Complete(solution->address(), shard_mask);
-      contract->Detach();
+
+      // TODO(AB): charge limit
+      FeeManager::TransactionDetails tx_details{
+          solution->address(), solution->address(), shard_mask, solution->contract_digest(),
+          CHARGE_RATE,         CHARGE_LIMIT,        false};
+
+      ContractExecutionResult result;
+      FETCH_LOG_INFO(LOGGING_NAME, "Verify: tx created");
+
+      {
+        telemetry::FunctionTimer const timer{*complete_duration_};
+        status = contract->Complete(
+            solution->address(), shard_mask, [this, &contract, &tx_details, &result]() -> bool {
+              return fee_manager_.CalculateChargeAndValidate(tx_details, {contract.get()}, result);
+            });
+      }
 
       if (SynergeticContract::Status::SUCCESS != status)
       {
         FETCH_LOG_WARN(LOGGING_NAME, "Failed to complete contract: 0x", contract->digest().ToHex(),
                        " Reason: ", ToString(status));
         return;
       }
+      FETCH_LOG_INFO(LOGGING_NAME, "Calculated fee: ", result.charge);
+      fee_manager_.Execute(tx_details, result, solution->block_index(), storage_);
```

Comment:
Maybe I am missing something but this deducts the fees from contract but I can't find where the fees are credited back